### DISCUSSION TITLE
Issue with AutoComplete of Function Signatures in Derived Classes

### DISCUSSION THREAD

**Discussion by glynor** [microsoft/pylance-release#4272] [General] [answered]
Let me preface with that it is entirely possible that this issue is "my fault" somehow, or that it is not an issue with Pylance but elsewhere in VS Code, but I figured I'd start here.

First, my details: I'm using Visual Studio Code 1.77.3 on macOS 12.6.5 on multiple machines (some Intel, some Apple Silicon). I'm using Python 3.11.3 currently, but the issue occurred with earlier point releases of 3.11.x as well. No virtual environment, as I'm not using _anything_ for these projects other than python standard library stuff. Pylance and Python Extensions are installed in VS Code in "stock" mode.

Version: 1.77.3 (Universal)
Commit: 704ed70d4fd1c6bd6342c436f1ede30d1cff4710
Date: 2023-04-12T09:19:37.325Z
Electron: 19.1.11
Chromium: 102.0.5005.196
Node.js: 16.14.2
V8: 10.2.154.26-electron.0
OS: Darwin x64 21.6.0
Sandboxed: Yes

Pylance: v2023.4.30
Python: v2023.6.1

[autocomplete_issue.py.zip](https://github.com/microsoft/pylance-release/files/11298056/autocomplete_issue.py.zip)

For the past few weeks (perhaps a month or even two, my memory is hazy and I ignored it for a while), I've been experiencing issues with Intellisense being inconsistent in its ability to auto-complete function signatures in derived classes when inheriting from a parent class. I thought it was perhaps some lazy typing, but I went through and made sure all of my code passed at least the basic typing tests in VS Code (without cheating with ignores and lazy casts) and it didn't help. So I've recently investigated further with a simple, self-contained test, and I can definitely reproduce weirdness on multiple machines.

Basically, I'm seeing Intellisense _sometimes_ fail to offer auto-complete for "overriding" a parent method signature. The issue seems to be triggered most reliably if you have a fairly large number functions in the base and derived class (though nothing astounding, 10 or 12 seems to do it). You can "bring back" the ability to override methods from the base class by removing some of your existing functions (as though there is a "limit" to how many functions you can have in a subclass). I can also reproduce it most reliably if the base and derived classes have methods using the property descriptor. And it seems to be impacted by whitespace and the order of the functions in the base and derived classes (even if they don't inter-relate in any way).

I've attached a simple example python file. It has a base class with a bunch of methods named base_method1, base_method2, etc. If I override a bunch of these in the derived class, eventually it "breaks" and then auto-complete starts acting oddly depending on where I insert my cursor to start defining a new function.

For example, in the attached file, if I position my cursor "within" the existing DerivedClass somewhere and try to override base_method7, 8, or 9 (which are in BaseClass, but not yet overridden in DerivedClass), it only finds the other methods I've already overriden:
![Screen Shot 2023-04-21 at 1 37 08 PM](https://user-images.githubusercontent.com/31105627/233700192-069755b3-582f-4e43-abce-b2bf7a45fb4c.png)

If I position my cursor at the very bottom of the file, I get them properly, but this time it doesn't offer ANY other choices:
![Screen Shot 2023-04-21 at 1 38 43 PM](https://user-images.githubusercontent.com/31105627/233700331-31f57926-77c3-499e-8ed5-75451a9914b5.png)

But even that only works if I make sure there's nothing of any kind (even whitespace) at all after the spot where I'm inserting:
![Screen Shot 2023-04-21 at 1 40 23 PM](https://user-images.githubusercontent.com/31105627/233700612-def7ddf7-6d7f-4b06-bf78-1604af47dab4.png)

**Comment by glynor** [comment]
Hopefully this provides enough detail for someone else to reproduce it. The inconsistent behavior of auto-complete seems to be frustratingly inconsistent. In my actual code, I have some derived classes that function perfectly normally, and then others that are broken (derived from the same base classes). Moving them around in the code, and removing or commenting out existing functions (but unrelated ones) changes how they behave with auto-complete when trying to add in new functions.

But my actual code is has interrelated classes and real functionality so I was glad when I could reproduce something similar in a very simple plain py file with no imports of any kind.

**Reply by debonte** [reply]
I'll take a look.

**Reply by debonte** [reply, marked as answer]
Thanks for the file and repro steps. I was able to repro the issue and filed https://github.com/microsoft/pylance-release/issues/4274.
